File: android/app/src/main/java/ai/oriveo/community/core/provider/LocalEngineConnector.kt
package ai.oriveo.community.core.provider

import ai.oriveo.community.core.model.RelayAuthMode
import ai.oriveo.community.core.model.RelayConnectionSecurityMode
import ai.oriveo.community.core.model.RelayRequestedConfig
import ai.oriveo.community.core.model.RelayTransport
import ai.oriveo.community.core.model.LocalModelLoadState
import ai.oriveo.community.core.model.LocalModelRuntimeMetadata
import ai.oriveo.community.core.model.ModelExecutionLocality
import ai.oriveo.community.core.model.ProviderServiceError
import ai.oriveo.community.core.provider.relay.RELAY_SECURITY_MODE_HEADER
import ai.oriveo.community.core.provider.relay.RELAY_CERTIFICATE_FINGERPRINT_HEADER
import ai.oriveo.community.core.provider.relay.relayLocalSecurityPlugin
import io.ktor.client.HttpClient
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.header
import io.ktor.client.request.setBody
import io.ktor.client.statement.bodyAsText
import io.ktor.http.HttpHeaders
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

data class LocalEngineConnection(
    val engine: LocalEngineKind,
    val endpoint: String,
    val apiBaseUrl: String,
    val modelIds: List<String>,
    val runtimeMetadata: Map<String, LocalModelRuntimeMetadata>,
    val selectedModelId: String,
    val requested: RelayRequestedConfig,
)

enum class LocalEngineConnectionFailure { InvalidEndpoint, CleartextCredentials, AuthenticationRejected, WrongEngine, Loading, NoModels, EngineStopped, OutOfMemory, ContextExceeded, Timeout, Network }

class LocalEngineConnectionException(val failure: LocalEngineConnectionFailure, cause: Throwable? = null) :
    IllegalStateException(failure.name, cause)

class LocalEngineConnector(
    client: HttpClient,
    private val json: Json,
    private val relayService: RelayService,
) {
    companion object {
        internal fun supportsSecurityMode(mode: RelayConnectionSecurityMode): Boolean = mode in setOf(
            RelayConnectionSecurityMode.RemoteHttps,
            RelayConnectionSecurityMode.LocalHttp,
            RelayConnectionSecurityMode.PrivateVpn,
            RelayConnectionSecurityMode.TofuHttps,
        )

        internal fun configurationFailure(error: Throwable): LocalEngineConnectionFailure =
            if ((error as? ProviderServiceError.InvalidConfiguration)?.detail == "cleartext_credentials") {
                LocalEngineConnectionFailure.CleartextCredentials
            } else {
                LocalEngineConnectionFailure.InvalidEndpoint
            }
    }

    private val client = client.config {
        followRedirects = false
        expectSuccess = false
        install(relayLocalSecurityPlugin())
    }

    suspend fun connect(
        engine: LocalEngineKind,
        rawEndpoint: String,
        securityMode: RelayConnectionSecurityMode,
        modelHint: String? = null,
        apiKey: String = "",
        certificateFingerprint: String? = null,
    ): LocalEngineConnection {
        if (!supportsSecurityMode(securityMode)) {
            throw LocalEngineConnectionException(LocalEngineConnectionFailure.InvalidEndpoint)
        }
        val requiresBearer = engine == LocalEngineKind.OpenWebUI
        if (requiresBearer && apiKey.isBlank()) throw LocalEngineConnectionException(LocalEngineConnectionFailure.InvalidEndpoint)
        val endpoint = runCatching {
            RelayEndpointPolicy.requireConfigured(
                rawEndpoint,
                securityMode,
                RelayEndpointPolicy.Credentials(authMode = if (requiresBearer) RelayAuthMode.Bearer else RelayAuthMode.None, hasKey = requiresBearer),
            )
        }.getOrElse { throw LocalEngineConnectionException(configurationFailure(it), it) }
        val template = LocalEngineContract.templates.getValue(engine)
        val requested = RelayRequestedConfig(
            transport = if (engine == LocalEngineKind.LlamaCpp) RelayTransport.LlamaCppNative else RelayTransport.OpenAIChatCompletions,
            authMode = if (requiresBearer) RelayAuthMode.Bearer else RelayAuthMode.None,
            securityMode = securityMode,
            stream = true,
            resolvedAPIBaseURL = apiBaseUrl(endpoint, engine),
            engineProfile = engine.wireName,
            certificateFingerprint = certificateFingerprint,
        )

        val fingerprint = get(endpoint, template.probePath, requested, apiKey)
        when (LocalEngineContract.classify(engine, fingerprint.status, fingerprint.contentType, fingerprint.json)) {
            LocalEngineState.Ready -> Unit
            LocalEngineState.Loading -> throw LocalEngineConnectionException(LocalEngineConnectionFailure.Loading)
            else -> throw LocalEngineConnectionException(LocalEngineConnectionFailure.WrongEngine)
        }
        val catalog = if (template.catalogPath == template.probePath) fingerprint
        else get(endpoint, template.catalogPath, requested, apiKey)
        val discovered = modelIds(engine, catalog.json)
        val runtimeMetadata = modelRuntimeMetadata(engine, catalog.json, discovered)
        val selected = modelHint?.trim()?.takeIf(String::isNotEmpty) ?: discovered.firstOrNull()
            ?: throw LocalEngineConnectionException(LocalEngineConnectionFailure.NoModels)

        val introspection = request(
            endpoint = endpoint,
            method = template.introspectionMethod,
            path = template.introspectionPath,
            body = if (engine == LocalEngineKind.Ollama) JsonObject(mapOf("model" to JsonPrimitive(selected))).toString() else null,
            requested = requested,
            apiKey = apiKey,
        )
        if (!introspectionMatches(engine, introspection)) {
            throw LocalEngineConnectionException(LocalEngineConnectionFailure.WrongEngine)
        }

        runCatching {
            relayService.pingRelay(apiKey, requested.resolvedAPIBaseURL, selected, requested)
        }.getOrElse { error ->
            if (error is LocalEngineConnectionException) throw error
            throw LocalEngineConnectionException(classifyNetworkFailure(error), error)
        }
        return LocalEngineConnection(
            engine = engine,
            endpoint = endpoint,
            apiBaseUrl = requested.resolvedAPIBaseURL ?: endpoint,
            modelIds = catalogModelIds(discovered, selected),
            runtimeMetadata = runtimeMetadata.ifEmpty {
                mapOf(selected to LocalModelRuntimeMetadata(LocalModelLoadState.Unknown, executionLocality(engine, selected)))
            },
            selectedModelId = selected,
            requested = requested,
        )
    }

    private suspend fun get(endpoint: String, path: String, requested: RelayRequestedConfig, apiKey: String): Snapshot {
        return request(endpoint, "GET", path, null, requested, apiKey)
    }

    private suspend fun request(endpoint: String, method: String, path: String, body: String?, requested: RelayRequestedConfig, apiKey: String): Snapshot {
        val url = endpoint.trimEnd('/') + "/" + path.trimStart('/')
        return runCatching {
            val response = if (method == "POST") client.post(url) {
                header(RELAY_SECURITY_MODE_HEADER, requested.securityMode.value)
                requested.certificateFingerprint?.let { header(RELAY_CERTIFICATE_FINGERPRINT_HEADER, it) }
                if (requested.authMode == RelayAuthMode.Bearer) header(HttpHeaders.Authorization, "Bearer $apiKey")
                header(HttpHeaders.ContentType, "application/json")
                if (body != null) setBody(body)
            } else client.get(url) {
                header(RELAY_SECURITY_MODE_HEADER, requested.securityMode.value)
                requested.certificateFingerprint?.let { header(RELAY_CERTIFICATE_FINGERPRINT_HEADER, it) }
                if (requested.authMode == RelayAuthMode.Bearer) header(HttpHeaders.Authorization, "Bearer $apiKey")
            }
            val body = response.bodyAsText()
            Snapshot(
                status = response.status.value,
                contentType = response.headers[HttpHeaders.ContentType].orEmpty(),
                json = runCatching { json.parseToJsonElement(body).jsonObject }.getOrNull(),
            ).also { snapshot ->
                if (snapshot.status == 401 || snapshot.status == 403) {
                    throw LocalEngineConnectionException(LocalEngineConnectionFailure.AuthenticationRejected)
                }
            }
        }.getOrElse { error ->
            if (error is LocalEngineConnectionException) throw error
            throw LocalEngineConnectionException(classifyNetworkFailure(error), error)
        }
    }

    private fun introspectionMatches(engine: LocalEngineKind, snapshot: Snapshot): Boolean {
        if (snapshot.status !in 200..299) return false
        if (engine == LocalEngineKind.Vllm) return true
        val body = snapshot.json ?: return false
        return when (engine) {
            LocalEngineKind.LlamaCpp -> body["default_generation_settings"] is JsonObject
            LocalEngineKind.Ollama -> body["capabilities"] != null || body["parameters"] != null
            LocalEngineKind.LmStudio -> body["data"] != null || body["models"] != null
            LocalEngineKind.Vllm -> true
            LocalEngineKind.OpenWebUI -> body["data"] != null || body["models"] != null
        }
    }

    private fun apiBaseUrl(endpoint: String, engine: LocalEngineKind): String = endpoint.trimEnd('/').let {
        if (engine == LocalEngineKind.LlamaCpp) return it
        if (engine == LocalEngineKind.OpenWebUI) return "$it/api"
        if (it.endsWith("/v1", ignoreCase = true)) it else "$it/v1"
    }

    private fun modelIds(engine: LocalEngineKind, body: JsonObject?): List<String> {
        val values = if (engine == LocalEngineKind.Ollama) {
            body?.get("models")?.jsonArray.orEmpty().mapNotNull { it.jsonObject["name"]?.jsonPrimitive?.contentOrNull }
        } else if (engine == LocalEngineKind.OpenWebUI) {
            (body?.get("data")?.jsonArray.orEmpty() + body?.get("models")?.jsonArray.orEmpty()).mapNotNull {
                it.jsonObject["id"]?.jsonPrimitive?.contentOrNull ?: it.jsonObject["name"]?.jsonPrimitive?.contentOrNull
            }
        } else {
            body?.get("data")?.jsonArray.orEmpty().mapNotNull { it.jsonObject["id"]?.jsonPrimitive?.contentOrNull }
        }
        return values.map(String::trim).filter(String::isNotEmpty).distinct()
    }

    private fun modelRuntimeMetadata(
        engine: LocalEngineKind,
        body: JsonObject?,
        modelIds: List<String>,
    ): Map<String, LocalModelRuntimeMetadata> {
        val rows = when (engine) {
            LocalEngineKind.Ollama -> body?.get("models")?.jsonArray.orEmpty()
            LocalEngineKind.OpenWebUI -> body?.get("data")?.jsonArray.orEmpty() + body?.get("models")?.jsonArray.orEmpty()
            else -> body?.get("data")?.jsonArray.orEmpty()
        }
        return modelIds.associateWith { modelId ->
            val row = rows.map { it.jsonObject }.firstOrNull {
                val rowId = if (engine == LocalEngineKind.Ollama) {
                    it["name"]?.jsonPrimitive?.contentOrNull
                } else {
                    it["id"]?.jsonPrimitive?.contentOrNull ?: it["name"]?.jsonPrimitive?.contentOrNull
                }
                rowId == modelId
            }
            val state = when (row?.get("state")?.jsonPrimitive?.contentOrNull?.lowercase()) {
                "loaded" -> LocalModelLoadState.Loaded
                "loading" -> LocalModelLoadState.Loading
                "unloaded" -> LocalModelLoadState.Unloaded
                else -> if (engine == LocalEngineKind.Vllm || engine == LocalEngineKind.OpenWebUI) LocalModelLoadState.Loaded else LocalModelLoadState.Unknown
            }
            LocalModelRuntimeMetadata(state, executionLocality(engine, modelId))
        }
    }

    private fun executionLocality(engine: LocalEngineKind, modelId: String): ModelExecutionLocality =
        if (LocalEngineContract.modelLocality(engine, modelId) == "cloud") ModelExecutionLocality.ProxiedCloud
        else ModelExecutionLocality.Local

    private fun classifyNetworkFailure(error: Throwable): LocalEngineConnectionFailure {
        val message = error.message.orEmpty().lowercase()
        return when {
            "401" in message || "403" in message || "unauthorized" in message || "forbidden" in message ->
                LocalEngineConnectionFailure.AuthenticationRejected
            "timed out" in message || "timeout" in message -> LocalEngineConnectionFailure.Timeout
            "out of memory" in message || "oom" in message -> LocalEngineConnectionFailure.OutOfMemory
            "context" in message && ("length" in message || "window" in message || "too long" in message) -> LocalEngineConnectionFailure.ContextExceeded
            "connect" in message || "refused" in message || "unreachable" in message -> LocalEngineConnectionFailure.EngineStopped
            else -> LocalEngineConnectionFailure.Network
        }
    }

    private val LocalEngineKind.wireName: String
        get() = when (this) {
            LocalEngineKind.LlamaCpp -> "llamacpp"
            LocalEngineKind.Ollama -> "ollama"
            LocalEngineKind.LmStudio -> "lmstudio"
            LocalEngineKind.Vllm -> "vllm"
            LocalEngineKind.OpenWebUI -> "openwebui"
        }

    internal fun catalogModelIds(discovered: List<String>, selected: String): List<String> =
        if (selected in discovered) discovered else listOf(selected) + discovered

    private data class Snapshot(val status: Int, val contentType: String, val json: JsonObject?)
}

File: ios/Oriveo/Oriveo/Core/Providers/LocalEngineConnector.swift
import Foundation

struct LocalEngineConnection: Sendable {
    let engine: LocalEngineKind
    let endpoint: String
    let apiBaseURL: String
    let modelIDs: [String]
    let runtimeMetadata: [String: LocalModelRuntimeMetadata]
    let selectedModelID: String
    let requested: RelayRequestedConfig
}

enum LocalEngineConnectionError: LocalizedError, Equatable {
    case invalidEndpoint
    case wrongEngine
    case engineLoading
    case noModels
    case engineStopped
    case outOfMemory
    case contextExceeded
    case timeout
    case localNetworkDenied
    case cleartextCredentials

    var errorDescription: String? {
        switch self {
        case .invalidEndpoint: return L10n.tr("Invalid local address", table: .providers)
        case .wrongEngine: return L10n.tr("The service at this address is not the selected engine.", table: .providers)
        case .engineLoading: return L10n.tr("The local model is still loading.", table: .providers)
        case .noModels: return L10n.tr("No local models were found.", table: .providers)
        case .engineStopped: return L10n.tr("The local engine is not running.", table: .providers)
        case .outOfMemory: return L10n.tr("The local engine ran out of memory.", table: .providers)
        case .contextExceeded: return L10n.tr("The message exceeds this model's context window.", table: .providers)
        case .timeout: return L10n.tr("The local model took too long to load.", table: .providers)
        case .localNetworkDenied: return L10n.tr("Allow local network access in Settings.", table: .providers)
        case .cleartextCredentials:
            return L10n.tr("An unencrypted connection can't carry a key. Change the address and connection type to Public HTTPS.", table: .providers)
        }
    }
}

enum LocalEngineConnector {
    static func connect(
        engine: LocalEngineKind,
        endpoint rawEndpoint: String,
        securityMode: RelayConnectionSecurityMode,
        modelHint: String? = nil,
        apiKey: String = "",
        certificateFingerprint: String? = nil,
        session: URLSession = .shared
    ) async throws -> LocalEngineConnection {
        guard supports(securityMode: securityMode) else {
            throw LocalEngineConnectionError.invalidEndpoint
        }
        let authPolicy = LocalEngineAuthenticationPolicy.policy(for: engine)
        let hasCredential = !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        try validateAuthenticationPolicy(
            authPolicy,
            securityMode: securityMode,
            hasCredential: hasCredential
        )
        let endpoint: String
        do {
            endpoint = try RelayEndpointPolicy.requireConfigured(
                rawEndpoint,
                securityMode: securityMode,
                credentials: .init(authMode: authPolicy.authMode, hasKey: hasCredential)
            )
        } catch ProviderServiceError.invalidConfiguration(let reason) where reason == "cleartext_credentials" {
            throw LocalEngineConnectionError.cleartextCredentials
        } catch {
            throw LocalEngineConnectionError.invalidEndpoint
        }
        guard let template = LocalEngineTemplate.all[engine] else {
            throw LocalEngineConnectionError.wrongEngine
        }
        let requested = RelayRequestedConfig(
            transport: engine == .llamacpp ? .llamacppNative : .openaiChatCompletions,
            authMode: authPolicy.authMode,
            securityMode: securityMode,
            stream: true,
            resolvedAPIBaseURL: apiBaseURL(for: endpoint, engine: engine),
            engineProfile: engine.rawValue,
            certificateFingerprint: certificateFingerprint
        )

        let fingerprint = try await get(
            endpoint: endpoint,
            path: template.probePath,
            requested: requested,
            apiKey: apiKey,
            session: session
        )
        switch LocalEngineContract.classify(
            engine: engine,
            status: fingerprint.status,
            contentType: fingerprint.contentType,
            json: fingerprint.json
        ) {
        case .ready: break
        case .loading: throw LocalEngineConnectionError.engineLoading
        default: throw LocalEngineConnectionError.wrongEngine
        }

        let catalog = template.catalogPath == template.probePath
            ? fingerprint
            : try await get(endpoint: endpoint, path: template.catalogPath, requested: requested, apiKey: apiKey, session: session)
        let discovered = modelIDs(engine: engine, json: catalog.json)
        let runtimeMetadata = modelRuntimeMetadata(engine: engine, json: catalog.json, modelIDs: discovered)
        let hinted = modelHint?.trimmingCharacters(in: .whitespacesAndNewlines)
        let selected = hinted.flatMap { $0.isEmpty ? nil : $0 } ?? discovered.first
        guard let selected else { throw LocalEngineConnectionError.noModels }

        let introspection = try await request(
            endpoint: endpoint,
            method: template.introspectionMethod,
            path: template.introspectionPath,
            jsonBody: engine == .ollama ? ["model": selected] : nil,
            requested: requested,
            apiKey: apiKey,
            session: session
        )
        guard introspectionMatches(engine: engine, snapshot: introspection) else {
            throw LocalEngineConnectionError.wrongEngine
        }

        do {
            _ = try await OpenAIService(session: session).pingRelay(
                apiKey: apiKey,
                baseURL: requested.resolvedAPIBaseURL ?? endpoint,
                modelID: selected,
                relayRequested: requested
            )
        } catch {
            throw classifyNetworkError(error)
        }
        return LocalEngineConnection(
            engine: engine,
            endpoint: endpoint,
            apiBaseURL: requested.resolvedAPIBaseURL ?? endpoint,
            modelIDs: catalogModelIDs(discovered: discovered, selected: selected),
            runtimeMetadata: runtimeMetadata.isEmpty
                ? [selected: .init(loadState: .unknown, executionLocality: executionLocality(engine: engine, modelID: selected))]
                : runtimeMetadata,
            selectedModelID: selected,
            requested: requested
        )
    }

    nonisolated static func catalogModelIDs(discovered: [String], selected: String) -> [String] {
        discovered.contains(selected) ? discovered : [selected] + discovered
    }

    /// Lightweight discovery proof. It verifies the selected engine's signature through the
    /// production security executor but deliberately does not claim generation success.
    static func verifyCandidate(
        engine: LocalEngineKind,
        endpoint rawEndpoint: String,
        securityMode: RelayConnectionSecurityMode,
        apiKey: String = "",
        session: URLSession = .shared
    ) async throws -> String {
        let authPolicy = LocalEngineAuthenticationPolicy.policy(for: engine)
        let hasCredential = !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        try validateAuthenticationPolicy(
            authPolicy,
            securityMode: securityMode,
            hasCredential: hasCredential
        )
        let endpoint: String
        do {
            endpoint = try RelayEndpointPolicy.requireConfigured(
                rawEndpoint,
                securityMode: securityMode,
                credentials: .init(authMode: authPolicy.authMode, hasKey: hasCredential)
            )
        } catch ProviderServiceError.invalidConfiguration(let reason) where reason == "cleartext_credentials" {
            throw LocalEngineConnectionError.cleartextCredentials
        } catch {
            throw LocalEngineConnectionError.invalidEndpoint
        }
        guard let template = LocalEngineTemplate.all[engine] else {
            throw LocalEngineConnectionError.wrongEngine
        }
        let requested = RelayRequestedConfig(
            transport: engine == .llamacpp ? .llamacppNative : .openaiChatCompletions,
            authMode: authPolicy.authMode,
            securityMode: securityMode,
            stream: true,
            resolvedAPIBaseURL: apiBaseURL(for: endpoint, engine: engine),
            engineProfile: engine.rawValue
        )
        let snapshot = try await get(
            endpoint: endpoint,
            path: template.probePath,
            requested: requested,
            apiKey: apiKey,
            session: session
        )
        switch LocalEngineContract.classify(
            engine: engine,
            status: snapshot.status,
            contentType: snapshot.contentType,
            json: snapshot.json
        ) {
        case .ready: return endpoint
        case .loading: throw LocalEngineConnectionError.engineLoading
        default: throw LocalEngineConnectionError.wrongEngine
        }
    }

    nonisolated static func supports(securityMode: RelayConnectionSecurityMode) -> Bool {
        switch securityMode {
        case .remoteHTTPS, .localHTTP, .privateVPN, .tofuHTTPS: return true
        }
    }

    private static func validateAuthenticationPolicy(
        _ policy: LocalEngineAuthenticationPolicy,
        securityMode: RelayConnectionSecurityMode,
        hasCredential: Bool
    ) throws {
        if policy.requiresCredential, hasCredential, securityMode == .localHTTP {
            throw LocalEngineConnectionError.cleartextCredentials
        }
        guard policy.permits(securityMode: securityMode, hasCredential: hasCredential) else {
            throw LocalEngineConnectionError.invalidEndpoint
        }
    }

    private static func apiBaseURL(for endpoint: String, engine: LocalEngineKind) -> String {
        let trimmed = endpoint.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        if engine == .llamacpp { return trimmed }
        if engine == .openwebui { return trimmed + "/api" }
        return trimmed.lowercased().hasSuffix("/v1") ? trimmed : "\(trimmed)/v1"
    }

    private static func get(
        endpoint: String,
        path: String,
        requested: RelayRequestedConfig,
        apiKey: String,
        session: URLSession
    ) async throws -> (status: Int, contentType: String, json: Any?) {
        try await request(endpoint: endpoint, method: "GET", path: path, jsonBody: nil, requested: requested, apiKey: apiKey, session: session)
    }

    private static func request(
        endpoint: String,
        method: String,
        path: String,
        jsonBody: [String: Any]?,
        requested: RelayRequestedConfig,
        apiKey: String,
        session: URLSession
    ) async throws -> (status: Int, contentType: String, json: Any?) {
        guard let base = URL(string: endpoint),
              let url = URL(string: path, relativeTo: base)?.absoluteURL else {
            throw LocalEngineConnectionError.invalidEndpoint
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 12
        if let jsonBody {
            request.httpBody = try JSONSerialization.data(withJSONObject: jsonBody)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        request.applyRelaySecurityMode(requested)
        if requested.authMode == .bearer { request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization") }
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.relayData(for: request)
        } catch {
            throw classifyNetworkError(error)
        }
        guard let http = response as? HTTPURLResponse else { throw LocalEngineConnectionError.wrongEngine }
        return (
            http.statusCode,
            http.value(forHTTPHeaderField: "Content-Type") ?? "",
            try? JSONSerialization.jsonObject(with: data)
        )
    }

    private static func introspectionMatches(
        engine: LocalEngineKind,
        snapshot: (status: Int, contentType: String, json: Any?)
    ) -> Bool {
        guard (200..<300).contains(snapshot.status) else { return false }
        if engine == .vllm { return true }
        guard snapshot.contentType.lowercased().contains("json"), let object = snapshot.json as? [String: Any] else { return false }
        switch engine {
        case .llamacpp: return object["default_generation_settings"] is [String: Any]
        case .ollama: return object["capabilities"] is [Any] || object["parameters"] is String
        case .lmstudio: return object["data"] is [[String: Any]] || object["models"] is [[String: Any]]
        case .vllm: return true
        case .openwebui: return object["data"] is [[String: Any]] || object["models"] is [[String: Any]]
        }
    }

    private static func modelIDs(engine: LocalEngineKind, json: Any?) -> [String] {
        guard let object = json as? [String: Any] else { return [] }
        let raw: [String]
        if engine == .ollama {
            raw = (object["models"] as? [[String: Any]] ?? []).compactMap { $0["name"] as? String }
        } else if engine == .openwebui {
            raw = (object["data"] as? [[String: Any]] ?? object["models"] as? [[String: Any]] ?? [])
                .compactMap { $0["id"] as? String ?? $0["name"] as? String }
        } else {
            raw = (object["data"] as? [[String: Any]] ?? []).compactMap { $0["id"] as? String }
        }
        return raw.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
            .reduce(into: []) { if !$0.contains($1) { $0.append($1) } }
    }

    private static func modelRuntimeMetadata(
        engine: LocalEngineKind,
        json: Any?,
        modelIDs: [String]
    ) -> [String: LocalModelRuntimeMetadata] {
        guard let object = json as? [String: Any] else { return [:] }
        let rows = engine == .ollama
            ? (object["models"] as? [[String: Any]] ?? [])
            : (object["data"] as? [[String: Any]] ?? [])
        return Dictionary(uniqueKeysWithValues: modelIDs.map { modelID in
            let row = rows.first { ($0[engine == .ollama ? "name" : "id"] as? String) == modelID }
            let rawState = (row?["state"] as? String)?.lowercased()
            let state: LocalModelLoadState = switch rawState {
            case "loaded": .loaded
            case "loading": .loading
            case "unloaded": .unloaded
            default: engine == .vllm || engine == .openwebui ? .loaded : .unknown
            }
            return (modelID, .init(
                loadState: state,
                executionLocality: executionLocality(engine: engine, modelID: modelID)
            ))
        })
    }

    private static func executionLocality(engine: LocalEngineKind, modelID: String) -> ModelExecutionLocality {
        LocalEngineContract.locality(engine: engine, modelID: modelID) == "cloud" ? .proxiedCloud : .local
    }

    static func classifyNetworkError(_ error: Error) -> LocalEngineConnectionError {
        if isLocalNetworkPermissionDenied(error) { return .localNetworkDenied }
        if let urlError = error as? URLError {
            if urlError.code == .timedOut { return .timeout }
            if [.cannotConnectToHost, .networkConnectionLost, .notConnectedToInternet, .cannotFindHost]
                .contains(urlError.code) { return .engineStopped }
        }
        let message = error.localizedDescription.lowercased()
        if message.contains("out of memory") || message.contains("oom") { return .outOfMemory }
        if message.contains("context") && (message.contains("length") || message.contains("window") || message.contains("too long")) {
            return .contextExceeded
        }
        return .wrongEngine
    }

    private static func isLocalNetworkPermissionDenied(_ error: Error) -> Bool {
        var current: NSError? = error as NSError
        for _ in 0..<6 {
            guard let candidate = current else { break }
            let diagnostic = ([candidate.localizedDescription] + candidate.userInfo.values.map { String(describing: $0) })
                .joined(separator: " ")
                .lowercased()
            if diagnostic.contains("local network prohibited") || diagnostic.contains("localnetworkdenied") {
                return true
            }
            current = candidate.userInfo[NSUnderlyingErrorKey] as? NSError
        }
        return false
    }
}
